wx._core.wxAssertionError: C++ assertion ""image.IsOk()"" failed when installed using pip

Searching online for this obscure-sounding error followed by "bitmap.cpp(593) in wxBitmap(): invalid image" in wx's C++ core didn't yield much except for other people confused by it, but on a closer look by a developer (me), it just means the image is missing. I'm stating the basics in case of other people finding this issue and getting confused by it. The issue is that the program is not looking in the right places for certain images, and I provide code below that solves it for cases where pronterface as installed via pip.

- Version: 2.0.0rc8 (via pip)
- OS: Devuan 4 (chimaera) (based on Debian 11 (bullseye))

*Note about not using a virtualenv: Yeah, I know everyone says use a virtualenv but I have several programs installed like this (to avoid old bugs in programs due to Debian and Devuan policies regarding vetting packages for apt), and automating updates like `cd Printrun && git pull && cd .. && python3 -m pip install --user --upgrade Printrun` stores everything in .local and I can prevent pip from having to compile wx (and several other dependencies) multiple times (for each virtualenv that uses it) when the dependency updates.*

## Actual behavior
The imagefile function in utils.py instead only looks in:
- candidate="/usr/share/pronterface/images"
- local_candidate="$HOME/.local/bin/images" (this one seems wrong)
- frozen_candidate="$HOME/.local/lib/python3.9/site-packages/printrun/images"
- relative="$HOME/.local/bin/images"


## Expected behavior
local_candidate seems like it should be "$HOME/.local/share/pronterface/images" instead of using the bin directory like `relative` already does, but if you really want to keep that for some reason (), that's fine (I use a separate `share_candidate` variable in the first added check below). This is confirmed to fix the error caused by the pip install then running pronterface.py from the path (from anywhere, using only the filename to run it). However, though the GUI now loads I still get the GUI error:\
09:28:46 AM: can't open file 'pronterface.png' (error 2: No such file or directory)\
09:28:46 AM: Failed to load image from file "pronterface.png".

Therefore, I also check in ./.local/share/pixmaps below.

In the imagefile function in utils.py, change
```Python
    local_candidate = os.path.join(os.path.dirname(sys.argv[0]),
                                   "images", filename)
    if os.path.exists(local_candidate):
        return local_candidate
```

to 
```Python
    local_candidate = os.path.join(os.path.dirname(sys.argv[0]),
                                   "images", filename)
    if os.path.exists(local_candidate):
        return local_candidate
    # The code above isn't changed. Below is new:
    my_local_share = os.path.join(
        os.path.dirname(os.path.dirname(sys.argv[0])),
        "share",
        "pronterface"
    )
    share_candidate = os.path.join(my_local_share, "images", filename)
    if os.path.exists(share_candidate):
        return share_candidate

```

Or if you prefer, replace the entire function with something neater, using `lookup_file`:
```Python
def imagefile(filename):
    my_local_share = os.path.join(
        os.path.dirname(os.path.dirname(sys.argv[0])),
        "share",
        "pronterface"
    )
    image_dirs = [
        os.path.join(DATADIR, 'pronterface', 'images'),
        os.path.join(os.path.dirname(sys.argv[0]), "images"),
        os.path.join(my_local_share, "images"),
        os.path.join(
            getattr(sys, "_MEIPASS", os.path.dirname(os.path.abspath(__file__))),
            "images"
        ),  # frozen
    ]
    path = lookup_file(filename, image_dirs)
    if path == filename:
        # The file wasn't found in any known location, so use a relative
        #   path.
        path = os.path.join("images", filename)
    return path
```

Both are tested and working.

In either case, you must also make the following change in the pixmapfile function in utils.py:

Change
```Python
    shared_pixmaps_dir = os.path.join(DATADIR, 'pixmaps')
    return lookup_file(filename, [shared_pixmaps_dir])
```

to
```Python
    shared_pixmaps_dir = os.path.join(DATADIR, 'pixmaps')
    local_pixmaps_dir = os.path.join(
        os.path.dirname(os.path.dirname(sys.argv[0])),
        "share",
        "pixmaps"
    )
    pixmaps_dirs = [shared_pixmaps_dir, local_pixmaps_dir]    
    return lookup_file(filename, pixmaps_dirs)
```

These two additional checks are confirmed to fix the error (occurs in several places if fixed in the caller rather than using the comprehensive fix above) caused by the pip install then running pronterface.py from the path (from anywhere, using only the filename to run it): `wx._core.wxAssertionError: C++ assertion ""image.IsOk()"" failed at /tmp/pip-install-0sdyo0e0/wxpython_e90d1c545c024e62b1f5e84d431bd0b3/ext/wxWidgets/src/gtk/bitmap.cpp(593) in wxBitmap(): invalid image`

Let me know if I should submit a PR and which version of the change you prefer for imagefile.

--- a/printrun/utils.py
+++ b/printrun/utils.py
@@ -86,27 +86,64 @@ def setup_logging(out, filepath = None, reset_handlers = False):
         logger.addHandler(logging_handler)
 
 def iconfile(filename):
+    '''
+    Look for an icon file in standard icon locations ("pixmaps"
+    directories) or use the windows_exe if applicable.
+    
+    Sequential arguments:
+    filename -- Provide a filename without the path.
+    
+    Returns:
+    The full path.
+    '''
     if hasattr(sys, "frozen") and sys.frozen == "windows_exe":
         return sys.executable
     else:
         return pixmapfile(filename)
 
 def imagefile(filename):
-    shared_pronterface_images_dir = os.path.join(DATADIR, 'pronterface/images')
-    candidate = os.path.join(shared_pronterface_images_dir, filename)
-    if os.path.exists(candidate):
-        return candidate
-    local_candidate = os.path.join(os.path.dirname(sys.argv[0]),
-                                   "images", filename)
-    if os.path.exists(local_candidate):
-        return local_candidate
-    frozen_candidate=os.path.join(getattr(sys, "_MEIPASS", os.path.dirname(os.path.abspath(__file__))),"images",filename)
-    if os.path.exists(frozen_candidate):
-        return frozen_candidate
-    else:
-        return os.path.join("images", filename)
+    '''
+    Get the full path to filename by checking standard image locations,
+    those being possible locations of the pronterface "images" directory
+    (See the lookup_file function's documentation for behavior).
+    '''
+    my_local_share = os.path.join(
+        os.path.dirname(os.path.dirname(sys.argv[0])),
+        "share",
+        "pronterface"
+    )
+    image_dirs = [
+        os.path.join(DATADIR, 'pronterface', 'images'),
+        os.path.join(os.path.dirname(sys.argv[0]), "images"),
+        os.path.join(my_local_share, "images"),
+        os.path.join(
+            getattr(sys, "_MEIPASS", os.path.dirname(os.path.abspath(__file__))),
+            "images"
+        ),  # INFO: lookup_file checks in frozen but not in frozen /images
+    ]
+    path = lookup_file(filename, image_dirs)
+    if path == filename:
+        # The file wasn't found in any known location, so use a relative
+        #   path.
+        path = os.path.join("images", filename)
+    return path
 
 def lookup_file(filename, prefixes):
+    '''
+    Look for an icon file in one or more prefixes in the given list,
+    or in the frozen data if applicable. If a result from this (or
+    from callers such as imagefile) is used for the wx.Image
+    constructor and filename doesn't exist as-is, the C++ part of wx
+    will raise an exception (wx._core.wxAssertionError): "invalid
+    image".
+    
+    Sequential arguments:
+    filename -- a filename without the path.
+    prefixes -- a list of paths.
+    
+    Returns:
+    The full path if found, or filename if not found.
+    '''
     local_candidate = os.path.join(os.path.dirname(sys.argv[0]), filename)
     if os.path.exists(local_candidate):
         return local_candidate
@@ -118,14 +155,34 @@ def lookup_file(filename, prefixes):
     return filename
 
 def pixmapfile(filename):
+    '''
+    Get the full path to filename by checking in standard icon
+    ("pixmaps") directories (See the lookup_file function's
+    documentation for behavior).
+    '''
     shared_pixmaps_dir = os.path.join(DATADIR, 'pixmaps')
-    return lookup_file(filename, [shared_pixmaps_dir])
+    local_pixmaps_dir = os.path.join(
+        os.path.dirname(os.path.dirname(sys.argv[0])),
+        "share",
+        "pixmaps"
+    )
+    pixmaps_dirs = [shared_pixmaps_dir, local_pixmaps_dir]
+    return lookup_file(filename, pixmaps_dirs)
 
 def sharedfile(filename):
+    '''
+    Get the full path to filename by checking in the shared
+    directory (See the lookup_file function's documentation for behavior).
+    '''
     shared_pronterface_dir = os.path.join(DATADIR, 'pronterface')
     return lookup_file(filename, [shared_pronterface_dir])
 
 def configfile(filename):
+    '''
+    Get the full path to filename by checking in the
+    standard configuration directory (See the lookup_file
+    function's documentation for behavior).
+    '''
     return lookup_file(filename, [os.path.expanduser("~/.printrun/"), ])
 
 def decode_utf8(s):

--- a/printrun/utils.py
+++ b/printrun/utils.py
@@ -130,10 +130,10 @@ def imagefile(filename):
 
 def lookup_file(filename, prefixes):
     '''
-    Look for an icon file in one or more prefixes in the given list,
-    or in the frozen data if applicable. If a result from this (or
-    from callers such as imagefile) is used for the wx.Image
-    constructor and filename doesn't exist as-is, the C++ part of wx
+    Get the full path to filename by checking one or more prefixes,
+    or in the frozen data if applicable. If a result from this
+    (or from callers such as imagefile) is used for the wx.Image
+    constructor and filename isn't found, the C++ part of wx
     will raise an exception (wx._core.wxAssertionError): "invalid
     image".
     

--- a/printrun/utils.py
+++ b/printrun/utils.py
@@ -106,7 +106,7 @@ def imagefile(filename):
         os.path.dirname(os.path.dirname(sys.argv[0])),
         "share",
         "pronterface"
-    )
+    )  # Used by pip install
     image_dirs = [
         os.path.join(DATADIR, 'pronterface', 'images'),
         os.path.join(os.path.dirname(sys.argv[0]), "images"),
@@ -114,7 +114,7 @@ def imagefile(filename):
         os.path.join(
             getattr(sys, "_MEIPASS", os.path.dirname(os.path.abspath(__file__))),
             "images"
-        ),  # INFO: lookup_file checks in frozen but not in frozen /images
+        ),  # Check manually since lookup_file checks in frozen but not /images
     ]
     path = lookup_file(filename, image_dirs)
     if path == filename:
@@ -160,7 +160,7 @@ def pixmapfile(filename):
         os.path.dirname(os.path.dirname(sys.argv[0])),
         "share",
         "pixmaps"
-    )
+    )  # Used by pip install
     pixmaps_dirs = [shared_pixmaps_dir, local_pixmaps_dir]
     return lookup_file(filename, pixmaps_dirs)
 

--- a/printrun/utils.py
+++ b/printrun/utils.py
@@ -87,8 +87,8 @@ def setup_logging(out, filepath = None, reset_handlers = False):
 
 def iconfile(filename):
     '''
-    Look for an icon file in standard icon locations ("pixmaps"
-    directories) or use the windows_exe if applicable.
+    Get the full path to filename by checking in standard icon locations
+    ("pixmaps" directories) or use the frozen executable if applicable.
     
     Sequential arguments:
     filename -- Provide a filename without the path.

--- a/printrun/utils.py
+++ b/printrun/utils.py
@@ -88,13 +88,8 @@ def setup_logging(out, filepath = None, reset_handlers = False):
 def iconfile(filename):
     '''
     Get the full path to filename by checking in standard icon locations
-    ("pixmaps" directories) or use the frozen executable if applicable.
-    
-    Sequential arguments:
-    filename -- Provide a filename without the path.
-    
-    Returns:
-    The full path.
+    ("pixmaps" directories) or use the frozen executable if applicable
+    (See the lookup_file function's documentation for behavior).
     '''
     if hasattr(sys, "frozen") and sys.frozen == "windows_exe":
         return sys.executable

